File: tree/leetcode_1448/solution.py
import pdb
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        res = [0]
        
        # pdb.set_trace()
        def dfs(node, max_node_num):
            if node.val >= max_node_num:
                res[0] += 1
            max_node_num = max(max_node_num, node.val)
            if node.left:
                dfs(node.left, max_node_num)
            if node.right:
                dfs(node.right, max_node_num)
        
        if root.left:
            dfs(root.left, root.val)
        
        if root.right:
            dfs(root.right, root.val)
        return res[0] + 1

File: tree/leetcode_1448/solution.rb
require "../builder"
# Definition for a binary tree node.

# @param {TreeNode} root
# @return {Integer}
def good_nodes(root)
    queue = [[root, root.val]]
    count = 0
    while queue.length > 0
        node, max = queue.shift
        count += 1 if node.val >= max
        queue.append([node.left, [node.left.val, max].max]) if node.left
        queue.append([node.right, [node.right.val, max].max]) if node.right
    end
    return count
end
test_cases = [
    {
        input: [3,1,4,3,nil,1,5],
        expected: 4
    },
    {
        input: [3,3,nil,4,2],
        expected: 3
    },
    {
        input: [1],
        expected: 1
    }
]

builder = BuilderTreeNode.new()
for test_case in test_cases
    root = builder.build(test_case[:input])
    result = good_nodes(root)

    if result != test_case[:expected]
        print("Not pass, result is #{result}\n")
    else
        print("Pass!\n")
    end
end